app_integrations/apps/salesforce.py

```diff
@@ -0,0 +1,395 @@
+        contain details about your org's user login history.
+
+    LoginAs events:
+        contain details about what a Salesforce admin did while logged in as another user.
+
+    Report events:
+        contain information about what happened when a user ran a report and
+        captures report's size.
+
+    ReportExport:
+        events contain details about reports that a user exported.
+    """
+    _SALESFORCE_TOKEN_URL = 'https://login.salesforce.com/services/oauth2/token'
+    _SALESFORCE_QUERY_URL = ('{instance_url}/services/data/v{api_version}/'
+                             '{query}{start_time}{event_type}')
+    # Use the Query resource to retrieve log files.
+    _SALESFORCE_QUERY_FILTERS = ('query?q=SELECT+Id+,+EventType+,+LogFile+,+LogDate+,'
+                                 '+LogFileLength+FROM+EventLogFile+')
+
+    # Log files filters. These typs of events should be auditted for security purpose.
+    _SALESFORCE_EVENT_TYPES = {'Console', 'Login', 'LoginAs', 'Report', 'ReportExport'}
+    # Use "Where" and "LogDate" clause to filter log files which are generated after
+    # last_timestamp.
+    _SALESFORCE_CREATE_AFTER = 'WHERE+LogDate+>+{}+'
+
+    _TIMEOUT = 15
+    EXCEPTIONS_TO_BACKOFF = (SalesforceAppError,)
+    BACKOFF_MAX_RETRIES = 3
+
+    def __init__(self, config):
+        super(SalesforceApp, self).__init__(config)
+        self._auth_headers = None
+        self._instance_url = None
+        self._latest_api_version = 0
+
+    @classmethod
+    def _type(cls):
+        return 'events'
+
+    @classmethod
+    def service(cls):
+        return 'salesforce'
+
+    @classmethod
+    def date_formatter(cls):
+        """Salesforce API date format: YYYY-MM-DDTHH:MM:SSZ"""
+        return '%Y-%m-%dT%H:%M:%SZ'
+
+    def _request_token(self):
+        """Request OAuth token frome salesforce
+
+        Meanwhile, it will also get instance url which will be used in futher
+        requests. The instance url identifies the Salesforce instance to which
+        API calls should be sent.
+
+        Returns:
+            bool: Returns True if update auth headers and instance url successfully.
+        """
+        headers = {'Content-Type': 'application/x-www-form-urlencoded'}
+        # required credentials when request for a token.
+        data = {
+            'grant_type': 'password',
+            'client_id': self._config.auth['client_id'],
+            'client_secret': self._config.auth['client_secret'],
+            'username': self._config.auth['username'],
+            'password': self._config.auth['password']+self._config.auth['security_token'],
+            'response_type': 'code',
+            'redirect_uri': self._SALESFORCE_TOKEN_URL
+        }
+        success, response = self._make_post_request(self._SALESFORCE_TOKEN_URL, headers, data)
+
+        if not success:
+            return False
+
+        if not response or not response.get('access_token') or not response.get('instance_url'):
```

Comment:
there's a lot of `not` negative logic here.. can we do this instead:

`if not (response and response.get('access_token') and response.get('instance_url'):...`

Reply:
@ryandeivert could we just use `all()` in this case?

Reply:
I will take @ryandeivert's suggestion. If use `all()`, it will check all conditions.
if use `if not (response and response.get('access_token') and response.get('instance_url')`, it will not check rest conditions if `response` if `None` or `False`.

Reply:
yeah I think `all()` is great for generators or iterables but not great when you're just passing in a bunch of vars